### PR TITLE
Consider storage_changed for assigning alias_of_input in aot_autograd when computing differentiable outputs that alias each other

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #115410
* #115497
* #112184
* __->__ #115315

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_functorch/_aot_autograd/collect_metadata_analysis.py [hunk 1]
```diff
@@ -303,7 +303,35 @@ def inner(*flat_args):
         # maps the id of an intermediate base to its index in the output of the compiled forward
         intermediate_base_tensor_id_to_output_idx: Dict[int, int] = {}
         intermediate_bases: List[torch.Tensor] = []
+        # Why do we care if storage changed?
+        # There is a really care class of situations, which basically only happen with something
```
Comment (voznesenskym) [inline torch/_functorch/_aot_autograd/collect_metadata_analysis.py:307]: care -> rare

**Comment by voznesenskym** [comment]
o needs a type check

**Review by bdhirsh — APPROVED** [review]
gonna spend a tiny bit time trying to come up with a potentially more reasonable `set_()` example, but stamping anyway (I think this is reasonable)

**Pull request closed without merging**